File: gitsearch.py
import requests
import datetime
import sys, os

version = "1.0"
site    = "deeplooklabs.com"
banner  = f"""

       _ _                       _    
  __ _(_) |_ ___ ___ __ _ _ _ __| |_  
 / _` | |  _(_-</ -_) _` | '_/ _| ' \ 
 \__, |_|\__/__/\___\__,_|_| \__|_||_|    {version}
 |___/                                

        {site}
"""

example = "python3 gitsearch.py \"tesla.com boto language:python\""

if len(sys.argv) < 2:
    print(banner)
    print(example)
    sys.exit()

search_term = sys.argv[1]
access_token = os.getenv('GITHUB_TOKEN')
sort_by = 'updated'
headers = {'Authorization': f'Token {access_token}'}
date_now = datetime.datetime.now()
year_now = str(date_now.year)
url = f'https://api.github.com/search/code?q={search_term}&sort={sort_by}'
response = requests.get(url, headers=headers)

def main():
    if response.status_code == 200:
        data = response.json()
        
        items = data.get('items', [])
        print(banner)
        print(f'[INF] Total found: {str(len(items))}')
        
        for item in items:
            file_name = item['name']
            file_url = item['html_url']
            
            file_response = requests.get(file_url, headers=headers)
            if file_response.status_code == 200:
                file_data = file_response.json()
                created_at = file_data['payload']['repo']['createdAt']
            else:
                created_at = 'Last update not found!'
            
            if created_at.startswith(year_now):
                print("[WRN] Recent result!")
                print(f'[{created_at}] [{file_name}] {file_url}')

            else:
                print(f'[{created_at}] [{file_name}] {file_url}')
    else:
        print('[WRN] An error occurred while making the request!')

if __name__ == '__main__':
    main()

File: cmd/gitsearch/gitsearch.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"strings"
	"time"
)

const (
	version = "1.0"
	site    = "deeplooklabs.com"
)

var (
	banner = fmt.Sprintf(`
       _ _                       _    
  __ _(_) |_ ___ ___ __ _ _ _ __| |_  
 / _`+"`"+` | |  _(_-</ -_) _`+"`"+` | '_/ _| ' \ 
 \__, |_|\__/__/\___\__,_|_| \__|_||_|    %s
 |___/                                

        %s
`, version, site)

	example = "go run gitsearch.go \"tesla.com boto language:python\""
)

type FileResponse struct {
	HTMLURL string `json:"html_url"`
	Payload struct {
		Repo struct {
			CreatedAt string `json:"created_at"`
		} `json:"repo"`
	} `json:"payload"`
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println(banner)
		fmt.Println(example)
		return
	}

	searchTerm := os.Args[1]
	accessToken := os.Getenv("GITHUB_TOKEN")
	sortBy := "updated"
	headers := map[string]string{"Authorization": "Token " + accessToken}
	dateNow := time.Now()
	yearNow := dateNow.Year()
	url := fmt.Sprintf("https://api.github.com/search/code?q=%s&sort=%s", searchTerm, sortBy)
	response, err := sendRequest(url, headers)
	if err != nil {
		fmt.Println("[WRN] An error occurred while making the request!")
		return
	}

	defer response.Body.Close()

	if response.StatusCode != 200 {
		fmt.Println("[WRN] An error occurred while making the request!")
		return
	}

	data, err := ioutil.ReadAll(response.Body)
	if err != nil {
		fmt.Println("[WRN] An error occurred while reading the response!")
		return
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		fmt.Println("[WRN] An error occurred while parsing the response!")
		return
	}

	items, ok := result["items"].([]interface{})
	if !ok {
		fmt.Println("[WRN] An error occurred while parsing the response!")
		return
	}

	fmt.Println(banner)
	fmt.Printf("[INF] Total found: %d\n", len(items))

	for _, item := range items {
		file, ok := item.(map[string]interface{})
		if !ok {
			fmt.Println("[WRN] An error occurred while parsing the response!")
			return
		}

		fileName, ok := file["name"].(string)
		if !ok {
			fmt.Println("[WRN] An error occurred while parsing the response!")
			return
		}

		fileURL, ok := file["html_url"].(string)
		if !ok {
			fmt.Println("[WRN] An error occurred while parsing the response!")
			return
		}

		fileResponse, err := sendRequest(fileURL, headers)
		if err != nil {
			fmt.Println("[WRN] An error occurred while making the request!")
			return
		}

		defer fileResponse.Body.Close()

		if fileResponse.StatusCode != 200 {
			fmt.Println("[WRN] An error occurred while making the request!")
			return
		}

		data, err := ioutil.ReadAll(fileResponse.Body)
		if err != nil {
			fmt.Println("[WRN] An error occurred while reading the response!")
			return
		}

		var fileData FileResponse
		if err := json.Unmarshal(data, &fileData); err != nil {
			fmt.Println("[WRN] An error occurred while parsing the response!")
			return
		}

		createdAt := fileData.Payload.Repo.CreatedAt
		if strings.HasPrefix(createdAt, fmt.Sprintf("%d", yearNow)) {
			fmt.Println("[WRN] Recent result!")
		}
		fmt.Printf("[%s] [%s] %s\n", createdAt, fileName, fileURL)
	}
}

func sendRequest(url string, headers map[string]string) (*http.Response, error) {
	client := &http.Client{}
	request, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	for key, value := range headers {
		request.Header.Set(key, value)
	}

	response, err := client.Do(request)
	if err != nil {
		return nil, err
	}

	return response, nil
}
